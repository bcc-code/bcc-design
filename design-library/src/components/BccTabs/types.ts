import { BccPin, BccBadge } from "@/index";
import type { VueComponent } from "@/types";

export type BccTabsGroup = Array<{
  title: string;
  icon?: VueComponent;
  pin?: InstanceType<typeof BccPin>["$props"];
  badge?: InstanceType<typeof BccBadge>["$props"] & { text: string };
  as?: VueComponent;
  disabled?: boolean;
  loading?: boolean;
}>;
