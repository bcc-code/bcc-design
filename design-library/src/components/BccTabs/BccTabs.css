@layer components {
    .bcc-tabs {
        @apply relative w-full overflow-hidden;
    }

    .bcc-tabs-bar {
        @apply relative flex items-end min-w-full overflow-scroll hide-scrollbar;
    }

    .bcc-tabs-bar::before {
        @apply absolute bottom-0 inset-x-0 border-b-2 border-on-primary;
        content: '';
    }

    @media (max-width: theme('screens.sm')) {
        .bcc-tabs-bar--1 .bcc-tabs-bar-item,        
        .bcc-tabs-bar--2 .bcc-tabs-bar-item {
            min-width: 50%;
        }

        .bcc-tabs-bar--3 .bcc-tabs-bar-item {
            min-width: 1/3;
        }

        .bcc-tabs-bar--4 .bcc-tabs-bar-item {
            min-width: 25%;
        }
    }

    .bcc-tabs-bar-item {
        @apply relative flex items-center justify-center whitespace-nowrap shrink-0 cursor-pointer select-none px-5 py-3 border-b-2 text-tertiary focus:outline-none transition-colors;
    }
    
    .bcc-tabs-bar-item:not(.bcc-tabs-bar-item--disabled) {
        @apply hover:text-button-tertiary-hover active:text-button-tertiary-active;
    }

    .bcc-tabs-bar-item--disabled {
        @apply opacity-60 cursor-not-allowed;
    }

    .bcc-tabs-bar-item--active {
        @apply border-button-primary-active text-button-tertiary-active;
    }

    .bcc-tabs-bar-icon {
        @apply w-5 h-5;
    }

    .bcc-tabs-item-title {
        @apply ml-1 mr-3;
    }

    /* Size */
    .bcc-tabs-sm {
        @apply px-3 py-2;
    }
    .bcc-tabs-lg {
        @apply px-6 py-4;
    }

    .bcc-tabs-sm .bcc-tabs-bar-icon {
        @apply w-4 h-4;
    }

    .bcc-tabs-lg .bcc-tabs-bar-icon {
        @apply w-6 h-6;
    }
}
