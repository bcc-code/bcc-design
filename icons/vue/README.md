# @bcc-code/icons-vue
Icons used in the BCC design system, built as Vue components.

[![version](https://img.shields.io/npm/v/@bcc-code/icons-vue)](https://github.com/bcc-code/bcc-icons/releases) [![license](https://img.shields.io/npm/l/@bcc-code/icons-vue)](https://github.com/bcc-code/bcc-icons/blob/main/LICENSE) [![install size](https://packagephobia.com/badge?p=@bcc-code/icons-vue)](https://packagephobia.com/result?p=@bcc-code/icons-vue)

## Installation & Usage
Refer to the [documentation](https://developer.bcc.no/bcc-design/icons) for installation instructions and usage information.

## License
This package is licensed under the [Apache 2.0 license](../LICENSE).
