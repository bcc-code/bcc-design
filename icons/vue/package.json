{
    "name": "@bcc-code/icons-vue",
    "repository": "https://github.com/bcc-code/bcc-icons.git",
    "homepage": "https://developer.bcc.no/bcc-design/icons",
    "version": "1.0.0",
    "license": "Apache-2.0",
    "files": [
        "dist"
    ],
    "main": "./dist/index.js",
    "module": "./dist/esm/index.js",
    "types": "dist/index.d.ts",
    "exports": {
        "import": "./dist/esm/index.js",
        "require": "./dist/index.js"
    },
    "peerDependencies": {
        "vue": ">= 3"
    }
}
