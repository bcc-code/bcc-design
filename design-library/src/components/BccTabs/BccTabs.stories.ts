import { AccountTreeIcon, CheckIcon, CircleNotificationsIcon } from "@bcc-code/icons-vue";
import BccTabs from "./BccTabs.vue";

import type { Meta, StoryFn } from "@storybook/vue3";
import { h } from "vue";

export default {
  title: "Components/BccTabs",
  component: BccTabs,
  argTypes: {
    tabs: {
      description: "Array of tabs you want to display",
    },
    size: {
      description: "The size of the tabs",
      options: ["sm", "base", "lg"],
      control: { type: "radio" },
    },
  },
} as Meta<typeof BccTabs>;

const Template: StoryFn<typeof BccTabs> = (args) => ({
  components: { BccTabs },
  setup() {
    return { args };
  },
  template: `<BccTabs v-bind="args"></BccTabs>`,
});

export const Example = Template.bind({});
Example.args = {
  tabs: [
    {
      title: "Using slots",
      icon: CircleNotificationsIcon,
      as: h({
        template: `<div class="p-4">
          <p>Define tab content using slots or components:</p>
          <code class="my-2 inline-block border-4 border-dashed border-gray-300 p-2 text-gray-600">
            &lt;template #tab-1&gt;You content here&lt;/template&gt;
          </code>
        </div>`,
      }),
    },
    {
      title: "Using Component",
      as: h({
        template: `
          <div class="p-4">
            <p>Or assign a component to the <code>as</code> field in tabs</p>
            <code class="my-2 inline-block border-4 border-dashed border-gray-300 p-2 text-gray-600">
              {
                title: "Features",
                as: YourComponent
              }
            </code>
          </div>
        `,
      }),
    },
    {
      title: "Disabled",
      disabled: true,
      badge: {
        text: "new",
        context: "info",
      },
    },
  ],
  size: "base",
};

Example.parameters = {
  docs: {
    source: {
      language: "html",
      code: `
<BccTabs :tabs="[
  {title:"Using slots", icon: CircleNotificationsIcon},
  {title:"Using Component", component: MyComponent},
  {title:"Disabled",disabled:true, badge: { text: "new", context: "info" }}
]'>
  <template #tab-1>
    <div class="p-4">This is tab 1</div>
  </template>

  <template #tab-3>
    <div class="p-4">This is tab 3</div>
  </template>
</BccTabs>
    `,
    },
  },
};

export const WithIconsPinsAndBadges = Template.bind({});
WithIconsPinsAndBadges.args = {
  tabs: [
    {
      title: "With Icon",
      icon: AccountTreeIcon,
      as: h({
        template: `<code class="my-2 inline-block border-4 border-dashed border-gray-300 p-2 text-gray-600">{
          title: "Features",
          icon: AccountTreeIcon,
        }</code>`,
      }),
    },
    {
      title: "With Pin",
      icon: CircleNotificationsIcon,
      pin: {
        text: "0/5",
        context: "info",
      },
      as: h({
        template: `<code class="my-2 inline-block border-4 border-dashed border-gray-300 p-2 text-gray-600">{
          title: "Features",
          pin: { text: "5" } },
        }</code>`,
      }),
    },
    {
      title: "With styled Pin",
      pin: {
        icon: CheckIcon,
        context: "success",
      },
      as: h({
        template: `<code class="my-2 inline-block border-4 border-dashed border-gray-300 p-2 text-gray-600">{
          title: "Features",
          pin: {
            icon: CheckIcon,
            context: "success",
          },
        }</code>`,
      }),
    },
    {
      title: "With Badge",
      badge: {
        text: "new",
        context: "info",
      },
      icon: AccountTreeIcon,
      as: h({
        template: `<code class="my-2 inline-block border-4 border-dashed border-gray-300 p-2 text-gray-600">{
          title: "Features",
          badge: {
            text: "new",
            context: "info",
          },
        }</code>`,
      }),
    },
  ],
  size: "base",
};

WithIconsPinsAndBadges.parameters = {
  docs: {
    source: {
      language: "html",
      code: `
<BccTabs :selectedIndex="selectedTab" @change="changeTab"
  :tabs="[
    {title:"Using slots", icon: AccountTreeIcon },
    {title:"With styled Pin", pin: { icon: CheckIcon, context: "success" } },
    {title:"With Badge", badge: { text: "new", context: "info"} }
  ]'></BccTabs>
    `,
    },
  },
};
