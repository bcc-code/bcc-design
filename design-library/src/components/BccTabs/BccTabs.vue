<script lang="ts">
export type BccTabsGroup = Array<{
  title: string;
  icon?: string | Component | Function;
  pin?: InstanceType<typeof BccPin>["$props"];
  badge?: InstanceType<typeof BccBadge>["$props"] & { text: string };
  as?: Component | Function;
  disabled?: boolean;
  loading?: boolean;
}>;
</script>

<script setup lang="ts">
import { BccBadge, BccPin } from "@/index";
import { Tab, TabGroup, TabList, TabPanel, TabPanels } from "@headlessui/vue";
import type { Component } from "vue";

type Props = {
  tabs: BccTabsGroup;
  size?: keyof typeof sizes;
};

withDefaults(defineProps<Props>(), {
  size: "base",
});

const sizes = {
  sm: "bcc-tabs-sm text-label-small",
  base: "text-label-base",
  lg: "bcc-tabs-lg text-label-medium",
};
</script>

<template>
  <TabGroup as="div" class="bcc-tabs">
    <TabList as="div" class="bcc-tabs-bar" :class="`bcc-tabs-bar--${tabs.length}`">
      <Tab
        v-for="(tab, index) in tabs"
        :disabled="!!tab.disabled"
        :key="index"
        v-slot="{ selected }"
        as="template"
      >
        <div
          class="bcc-tabs-bar-item"
          :class="[
            {
              'bcc-tabs-bar-item--active': selected,
              'bcc-tabs-bar-item--disabled': !!tab.disabled,
            },
            sizes[size],
          ]"
        >
          <component v-if="tab.icon" :is="tab.icon" class="bcc-tabs-bar-icon" />
          <span class="bcc-tabs-item-title">{{ tab.title }}</span>
          <BccPin v-if="tab.pin" v-bind="tab.pin" />
          <BccBadge v-if="tab.badge" v-bind="tab.badge">
            {{ tab.badge.text }}
          </BccBadge>
        </div>
      </Tab>
    </TabList>
    <TabPanels>
      <TabPanel v-for="(tab, index) in tabs" :key="`tab-${index + 1}`">
        <slot :name="`tab-${index + 1}`" :tab="tab">
          <component v-if="tab.as" :is="tab.as" />
        </slot>
      </TabPanel>
    </TabPanels>
  </TabGroup>
</template>
