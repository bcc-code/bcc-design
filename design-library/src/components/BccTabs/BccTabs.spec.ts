import { describe, expect, it } from "vitest";

import { CircleIcon } from "@bcc-code/icons-vue";
import { mount } from "@vue/test-utils";
import { h } from "vue";
import BccTabs from "./BccTabs.vue";

describe("BccTabs", () => {
  it("renders three tabs", () => {
    expect(BccTabs).toBeTruthy();

    const wrapper = mount(BccTabs, {
      props: {
        tabs: [
          {
            title: "With icon",
            icon: CircleIcon,
          },
          {
            title: "Using Component",
            pin: {
              text: "5",
            },
            as: h({
              template: `
                <div class="p-4">
                  <p>Component</p>
                </div>
              `,
            }),
          },
          {
            title: "Disabled badge",
            disabled: true,
            badge: {
              text: "new",
              context: "info",
            },
          },
        ],
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
