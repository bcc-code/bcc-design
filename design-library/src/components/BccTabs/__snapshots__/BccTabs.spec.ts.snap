// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BccTabs > renders three tabs 1`] = `
"<button aria-hidden=\\"true\\" style=\\"position: fixed; height: 0px; padding: 0px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;\\" type=\\"button\\"></button>
<div class=\\"bcc-tabs\\">
  <div role=\\"tablist\\" aria-orientation=\\"horizontal\\" class=\\"bcc-tabs-bar bcc-tabs-bar--3\\">
    <div class=\\"bcc-tabs-bar-item bcc-tabs-bar-item--active text-label-base\\" id=\\"headlessui-tabs-tab-1\\" role=\\"tab\\" aria-selected=\\"true\\" tabindex=\\"0\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" aria-hidden=\\"true\\" viewBox=\\"0 96 960 960\\" fill=\\"currentColor\\" class=\\"bcc-tabs-bar-icon\\">
        <path d=\\"M480.033 981.978q-83.468 0-157.541-31.878-74.073-31.879-129.393-87.199-55.32-55.32-87.199-129.36-31.878-74.04-31.878-157.508 0-84.468 31.878-158.541 31.879-74.073 87.161-128.906 55.283-54.832 129.341-86.818 74.057-31.986 157.545-31.986 84.488 0 158.589 31.968 74.102 31.967 128.916 86.768 54.815 54.801 86.79 128.883Q886.218 491.484 886.218 576q0 83.501-31.986 157.57-31.986 74.069-86.818 129.36-54.833 55.291-128.873 87.17-74.04 31.878-158.508 31.878Zm-.033-68.13q141.043 0 239.446-98.752Q817.848 716.344 817.848 576q0-141.043-98.402-239.446-98.403-98.402-239.566-98.402-140.163 0-238.945 98.402-98.783 98.403-98.783 239.566 0 140.163 98.752 238.945Q339.656 913.848 480 913.848ZM480 576Z\\"></path>
      </svg> With icon
      <!--v-if-->
      <!--v-if-->
    </div>
    <div class=\\"bcc-tabs-bar-item text-label-base\\" id=\\"headlessui-tabs-tab-2\\" role=\\"tab\\" aria-selected=\\"false\\" tabindex=\\"-1\\">
      <!--v-if--> Using Component <div class=\\"bcc-pin bcc-pin-text\\"><span>5</span></div>
      <!--v-if-->
    </div>
    <div class=\\"bcc-tabs-bar-item bcc-tabs-bar-item--disabled text-label-base\\" disabled=\\"true\\" id=\\"headlessui-tabs-tab-3\\" role=\\"tab\\" aria-selected=\\"false\\" tabindex=\\"-1\\">
      <!--v-if--> Disabled badge
      <!--v-if-->
      <div class=\\"bcc-badge bcc-badge-info\\" text=\\"new\\">
        <!--v-if--><span class=\\"order-2\\">new</span>
      </div>
    </div>
  </div>
  <div>
    <div id=\\"headlessui-tabs-panel-4\\" role=\\"tabpanel\\" tabindex=\\"0\\" data-headlessui-state=\\"selected\\">
      <!--v-if-->
    </div><span style=\\"position: fixed; height: 0px; padding: 0px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;\\" id=\\"headlessui-tabs-panel-5\\" role=\\"tabpanel\\" tabindex=\\"-1\\"></span><span style=\\"position: fixed; height: 0px; padding: 0px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;\\" id=\\"headlessui-tabs-panel-6\\" role=\\"tabpanel\\" tabindex=\\"-1\\"></span>
  </div>
</div>"
`;
